python/lsst/pipe/base/butlerQuantumContext.py

```diff
@@ -0,0 +1,204 @@
+            a single `lsst.daf.butler.DatasetRef`. The function will get and return
+            the corresponding datasets from the butler.
+
+        Raises
+        ------
+        ValueError
+            If a `DatasetRef` is passed to put that is not defined in the quantum object, or
+            the type of values does not match what is expected from the type of dataset.
+        """
+        if isinstance(dataset, OutputQuantizedConnection):
+            if not isinstance(values, Struct):
+                raise ValueError("dataset is a OutputQuantizedConnection, a Struct with corresponding"
+                                 "attributes must be passed as the values to put")
+            for name, refs in dataset:
+                valuesAttribute = getattr(values, name)
+                if isinstance(refs, list):
+                    if len(refs) != len(valuesAttribute):
+                        raise ValueError(f"There must be a object to put for every Dataset ref in {name}")
+                    for i, ref in enumerate(refs):
+                        self._put(valuesAttribute[i], ref)
+                else:
+                    self._put(valuesAttribute, refs)
+        elif isinstance(dataset, list):
+            if len(dataset) != len(values):
+                raise ValueError("There must be a common number of references and values to put")
+            for i, ref in enumerate(dataset):
+                self._put(values[i], ref)
+        elif isinstance(dataset, DatasetRef):
+            self._put(values, dataset)
+        else:
+            raise TypeError("Dataset argument is not a type that can be used to put")
+
+    def _checkMembership(self, ref: typing.Union[typing.List[DatasetRef], DatasetRef], inout: RefDirection):
+        """Internal function used to check if a DatasetRef is part of the input quantum
+
+        This function will raise an exception if the ButlerQuantumContext is used to
+        get/put a DatasetRef which is not defined in the quantum.
+
+        Parameters
+        ----------
+        ref : `list` of `DatasetRef` or `DatasetRef`
+            Either a list or a single `DatasetRef` to check
+        inout : `RefDirection` enum
+            The connection type to check, e.g. either an input or an output. This prevents
+            both types needing to be checked for every operation, which may be important
+            for Quanta with lots of `DatasetRef`s.
+        """
+        if not isinstance(ref, list):
+            ref = [ref]
+        if inout is RefDirection.IN:
+            for r in ref:
+                if (r.datasetType, r.dataId) not in self.allInputs:
+                    raise ValueError("DatasetRef is not part of the Quantum being processed")
+        if inout is RefDirection.OUT:
+            for r in ref:
+                if (r.datasetType, r.dataId) not in self.allOutputs:
+                    raise ValueError("DatasetRef is not part of the Quantum being processed")
```

Comment:
There is repetition in this method, I think you can simplify it if you just pass `self.allInputs` or `self.allOutputs` as a parameter instead of enum (and enum will not even have to exist in that case).

Reply:
Really good point, I changed this so many times I could not see the forest for the trees